### PR TITLE
logger: Add native backend for native_posix

### PR DESCRIPTION
**Pull request by finikorg**
Add backend to be used in native_posix arch.

### CODE REVIEW

**Comment by aescolar** [comment]
Oh well, you overtook me :)

File: subsys/logging/log_backend_native.c
Comment (aescolar) [inline subsys/logging/log_backend_native.c]: Just a note: In principle this is ok, but we need to check if stdout is fed to a file or console.
When fed to a file we should disable it, otherwise all the ascii escapes will be quite annoying.

File: subsys/logging/CMakeLists.txt
Comment (nordic-krch) [inline subsys/logging/CMakeLists.txt]: native of which country? ;) to user who do not care about native posix shortcut 'native' does not tell anything thus imo this should rather be log_backend_native_posix.

File: subsys/logging/log_backend_native.c
Comment (nordic-krch) [inline subsys/logging/log_backend_native.c]: reuse of uart backend option (CONFIG_LOG_BACKEND_UART_SHOW_COLOR) probable copy-paste error.

File: subsys/logging/log_backend_native.c
Comment (nordic-krch) [reply subsys/logging/log_backend_native.c]: yes, so this is a copy-paste of uart backend which is very simple one (polling) and might be reworked in the future. Here, it may make sense to have the buffer bigger if block operations are faster.

File: subsys/logging/log_backend_native.c
Comment (nordic-krch) [inline subsys/logging/log_backend_native.c]: are we sure that standard console will always work in panic (e.g. hardfault, error handler?) or it is possible that it relies on interrupt and cannot be used or need to be reconfigured once panic occurs? UART in polling mode will work in that case so panic mode does nothing.

Comment (aescolar) [reply subsys/logging/log_backend_native.c]: In native_posix yes.

**Comment by aescolar** [comment]
@finikorg 
Take a look at this branch:
https://github.com/aescolar/zephyr/tree/new_logger
In kind of works.. There is something very fishy happening in the logger itself, which I would need to debug.. I hope I can take a look at it tomorrow.

**Comment by aescolar** [comment]
@nordic-krch : Seems this function does not work when the length of the backend buffer is more than 1:
https://github.com/zephyrproject-rtos/zephyr/blob/master/subsys/logging/log_output.c#L55
Offset grows by 1 each time that function is called and the logger prints its output into the previous string..

**Comment by aescolar** [comment]
@nordic-krch : Please take a look at the hacks I did in the 2 commits in the core logger code in this branch:
https://github.com/aescolar/zephyr/tree/new_logger

File: subsys/logging/log_output.c
Comment (finikorg) [reply subsys/logging/log_output.c]: Otherwise it does not build because of _prf().

File: subsys/logging/log_backend_native.c
Comment (finikorg) [reply subsys/logging/log_backend_native.c]: I am thinking to rename `CONFIG_LOG_BACKEND_UART_SHOW_COLOR` to `CONFIG_LOG_SHOW_COLOR` so it might be used in both backends

**Comment by aescolar** [comment]
@finikorg : 
It is unfortunate that we both started working on this at the same time. 
(From your other PR comment, I understood you were blocked and waiting for me to add the backend, so I had started on it.)
After I saw yesterday that this PR was not working and had a DNM I just merged it into my branch.
Please look in this branch
https://github.com/aescolar/zephyr/tree/newlogger2
You can take it straight
That branch seems to work, and from my point of view is ready to merge.
It's up to you if you want to use it or you prefer to go with your own changes.

**Comment by finikorg** [comment]
@aescolar I tried to cherry-pick as much your patches as I can, only last one is missing at the moment. I needed this backend for my PR #9877 .

File: subsys/logging/Kconfig
Comment (aescolar) [inline subsys/logging/Kconfig]: Having here the `depends on ARCH_POSIX` would be nice

File: subsys/logging/log_backend_native_posix.c
Comment (aescolar) [inline subsys/logging/log_backend_native_posix.c]: I think it is worth while having the comment about why this is just 1 (by now), as otherwise it is difficult to explain why `buf` is 256 char long

File: subsys/logging/log_backend_native_posix.c
Comment (aescolar) [inline subsys/logging/log_backend_native_posix.c]: Could you add this flag only `if (posix_trace_over_tty(0))`?
That way users will still be able to disable it on runtime, and it will be disabled when piping to a file or when run from sanitycheck.

File: subsys/logging/log_backend_native_posix.c
Comment (aescolar) [inline subsys/logging/log_backend_native_posix.c]: The comment about why panic does not need to do anything would help other people understand why this can just be stub.
(for ex. @nordic-krch just asked that before)

File: subsys/logging/log_core.c
Comment (aescolar) [inline subsys/logging/log_core.c]: This may be a bit minor:
Each of this is an unused pointer when its backend is not configured in. I would think a `#define native_backend NULL` would be a better choice.

**Review by aescolar — DISMISSED** [review]
Just a few minor things. 
The only reason for the reject is the color comment.

**Comment by aescolar** [comment]
About the 2 failures in CI, they are both due to the new assert that checks that there must be a backend:
https://github.com/zephyrproject-rtos/zephyr/pull/9968/commits/fb0e038c75565b40550bcdf495cd53570a581b0f
That check ensures that the logger thread does not go into an infinite loop while doing nothing.

Both failures are the same, and for xtensa:
```
***** Booting Zephyr OS zephyr-v1.13.0-113-g79cdad94d *****
ASSERTION FAIL [log_backend_count_get() > 0] @ /home/buildslave/src/github.com/zephyrproject-rtos/zephyr/subsys/logging/log_core.c:504:
```
It seems these tests were passing before, because they do not actually require a backend. Now they do not because as soon as the assert is triggered that logger thread is killed.

@finikorg : I guess you could remove that assert as it is not the point of this PR to fix issues in the logger itself. But @nordic-krch : maybe you want to consider having that assert, and looking into the issue for this platform(?)

**Review by aescolar — COMMENTED** [review]
Ok by me.
Minor: You may want to change the commit messages for
* "native_posix: Add logger configuration"
(The description should probably say what the commit does. You can just add my signoff to it instead of mentioning in the message you got the change from my branch)
* "logger: Remove assert() in thread start" Could probably use a longer explanation of why that is being done (OR you can just squash that commit together into the commit where I added it, and save yourself from explaining why you remove it)

**Review by dbkinder — APPROVED** [review]
+1 for doc changes

**Comment by aescolar** [comment]
@finikorg Note that the changes to the logger merged in master are conflicting with this changes.

**Comment by finikorg** [comment]
@aescolar rebased

**Comment by aescolar** [comment]
@nordic-krch : are you happy with this new backend and with the piggybacked bugfixes in this PR?

File: subsys/logging/log_core.c
Comment (aescolar) [inline subsys/logging/log_core.c]: +1 for not having the different backends code in the common logger files.

**Review by aescolar — APPROVED** [review]
+2 to the backend
+1 for the refactoring in the logger code and bugfixes

File: subsys/logging/log_core.c
Comment (nordic-krch) [inline subsys/logging/log_core.c]: There will be use cases where we don't want to enable all backends. Example: log backend over bluetooth where backend will be registered but not activated until connection is established.

File: samples/subsys/logging/logger/src/main.c
Comment (aescolar) [reply samples/subsys/logging/logger/src/main.c]: This app was hanging in CI when run in native_posix. 
This is due to the way the posix arch works: *You can imagine the code executes in a simulated CPU which runs at an infinitely fast clock: No time passes while the CPU is running.*
https://docs.zephyrproject.org/latest/boards/posix/native_posix/doc/board.html#important-limitations

File: subsys/logging/log_core.c
Comment (aescolar) [reply subsys/logging/log_core.c]: Is the expectation then, that backends which are meant to be started if configured, be called from here (like the UART one was, but the others be activated from other places?

Comment (nordic-krch) [reply subsys/logging/log_core.c]: yes, i was thinking the same that we might want to extend backend registeration macro to has something autostart flag.

Comment (nordic-krch) [reply subsys/logging/log_core.c]: Lets keep this change as is and i will try to address that in separate PR.

Comment (aescolar) [reply subsys/logging/log_core.c]: > Lets keep this change as is and i will try to address that in separate PR.

That would be best yes.

>  extend backend registeration macro to has something autostart flag.

Note that would use memory, while keeping that information implicitly in the order would not.

Comment (finikorg) [reply subsys/logging/log_core.c]: I think we might handle some special cases in api->init()

**Comment by aescolar** [comment]
> Seems this function does not work when the length of the backend buffer is more than 1

@nordic-krch : Was this fixed? At least if I try now to increase the registered output size things seem to work ok.

**Comment by aescolar** [comment]
@nordic-krch : Is this PR good enough to get your approval?

**Comment by nordic-krch** [comment]
@aescolar 
> Was this fixed?

yes

>  Is this PR good enough to get your approval?

yes

**Comment by aescolar** [comment]
@carlescufi / @nashif Could either of you do a quick review, and if you are happy with it merge?

**Comment by carlescufi** [comment]
@aescolar I'm fine with this but would like @nordic-krch to approve before we merge.

**Comment by aescolar** [comment]
He has on the comment https://github.com/zephyrproject-rtos/zephyr/pull/9968#issuecomment-422361401

**Comment by aescolar** [comment]
@carlescufi : He has on the comment just above https://github.com/zephyrproject-rtos/zephyr/pull/9968#issuecomment-422361401

**Pull request merged**
